samples: bluetooth: ble_hrs_peripheral_central: Add sample (#721)
nrfconnect/sdk-nrf-bm: samples/bluetooth/ble_hrs_peripheral_central/README.rst

.. _ble_hrs_peripheral_central_sample:

Bluetooth: Heart Rate Service Peripheral Central
################################################

.. contents::
   :local:
   :depth: 2

The Heart Rate Service Peripheral Central sample demonstrates how you can implement the Heart Rate profile as a server and client using |BMlong|.

Requirements
************

The sample supports the following development kits:

.. tabs::

> 

Reviewer: We have a template that can be used here, see other central samples.

           - PCA10156
           - S145
           - bm_nrf54l15dk/nrf54l10/cpuapp/s145_softdevice/mcuboot
         * - `nRF54L15 DK`_ (emulating nRF54L05)
           - PCA10156
           - S145
           - bm_nrf54l15dk/nrf54l05/cpuapp/s145_softdevice/mcuboot

Overview
********

This sample advertises and scans for devices that advertise with the :ref:`lib_ble_service_hrs` UUID (0x180D) and initiates a connection when a device is found.
When a peripheral device is connected, the sample starts the service discovery procedure.
If this succeeds, the sample subscribes to the Heart Rate Measurement characteristic to receive heart rate notifications.
If a client connects to this device the client subscribes to the Heart Rate Measurement characteristic to receive heart rate notifications forwarded from the peripheral device.

.. _ble_hrs_peripheral_central_sample_testing:

User interface
**************

Button 0:
  Press to disable allow list.

Button 1:
  Press to disconnect from the connected peripheral device.

  Keep the button pressed while resetting the board to delete bonding information for all peers stored on the device.

Button 2:

>   Press to disconnect from the connected client device.

Reviewer: client and server, or peripheral and central.
```suggestion
  Press to disconnect from the connected central device.
```

>   Keep the button pressed while resetting the board to delete bonding information for all peers stored on the device.

Reviewer: This is covered by button 1(?)

LED 0:
   Lit when the device is initialized.

LED 1:
   Lit when a peripheral device is connected.

LED 2:
   Lit when a client device is connected.

Building and running
********************

This sample can be found under :file:`samples/bluetooth/ble_hrs_peripheral_central/` in the |BMshort| folder structure.

For details on how to create, configure, and program a sample, see :ref:`getting_started_with_the_samples`.

Testing
=======

> This sample requires three devices to test, one running this sample, another one running the :ref:`ble_hrs_sample` sample, and another running the :ref:`ble_hrs_central_sample`.

Reviewer: Or, as many might want, a phone acting as the second central

Complete the following steps to test the sample:

1. Compile and program the application.

> #. In the Serial Terminal, observe that the ``Advertising as nRF_BM_HRS_Hub`` message is printed.

Reviewer: Not sure if Bridge is the sample name we want to use for it, though this should match the name of the sample (if using bridge, update peripheral_central -> bridge in folder name, documentation etc.) 
```suggestion
#. In the Serial Terminal, observe that the ``Advertising as nRF_BM_HRS_Bridge`` message is printed.
```

   You can configure the advertising name using the :kconfig:option:`CONFIG_SAMPLE_BLE_DEVICE_NAME` Kconfig option.
   For information on how to do this, see `Configuring Kconfig`_.

> #. Observe that the ``BLE HRS Peripheral Central sample initialized`` message is printed.

Reviewer: We should put this log before the advertising log, swap order in the source file as well.

#. Program the second development kit with the :ref:`ble_hrs_sample` sample.
#. Program the third development kit with the :ref:`ble_hrs_central_sample` sample, set the :kconfig:option:`CONFIG_SAMPLE_TARGET_PERIPHERAL_NAME` Kconfig option to ``nRF_BM_HRS_Peripheral_Central``, and set the :kconfig:option:`CONFIG_SAMPLE_USE_TARGET_PERIPHERAL_NAME` Kconfig option to ``y``.

> #. Observe that the ``Scan filter match`` message is printed, followed by ``Connecting to target`` and ``Connected``, this will happen twice, once for each peer.

Reviewer: We should have only one scan filter match(?). The other central device is not advertising.

.. _ble_hrs_peripheral_central_sample:

Bluetooth: Heart Rate Service Peripheral Central
################################################

.. contents::
   :local:
   :depth: 2

The Heart Rate Service Peripheral Central sample demonstrates how you can implement the Heart Rate profile as a server and client using |BMlong|.

Requirements
************

The sample supports the following development kits:

.. tabs::

   .. group-tab:: Simple board variants

      The following board variants do **not** have DFU capabilities:

      .. include:: /includes/supported_boards_all_non-mcuboot_variants_s145.txt

   .. group-tab:: MCUboot board variants

      The following board variants have DFU capabilities:

      .. include:: /includes/supported_boards_all_mcuboot_variants_s145.txt

Overview
********

This sample advertises and scans for devices that advertise with the :ref:`lib_ble_service_hrs` UUID (0x180D) and initiates a connection when a device is found.
When a peripheral device is connected, the sample starts the service discovery procedure.
If this succeeds, the sample subscribes to the Heart Rate Measurement characteristic to receive heart rate notifications.

> If a central connects to this device the central subscribes to the Heart Rate Measurement characteristic to receive heart rate notifications forwarded from the peripheral device.

Reviewer: I think this could use a rewrite. Perhaps something like. 

```suggestion
This sample acts simultaneously as both a peripheral and central device. 
* As a peripheral the sample advertises with the :ref:`lib_ble_service_hrs` UUID (0x180D). If a central connects to this device the central subscribes to the Heart Rate Measurement characteristic to receive heart rate notifications forwarded from the peripheral device.
* As a central the sample scans for other devices that advertise with the :ref:`lib_ble_service_hrs` UUID (0x180D). When a device is found it initiates a connection before starting the service discovery procedure and subscribing to the Heart Rate Measurement characteristic notifications.
```

Reviewer 2: Rewrote. Please have another look.

> If a central connects to this device the central subscribes to the Heart Rate Measurement characteristic to receive heart rate notifications forwarded from the peripheral device.

Reviewer 3: ```suggestion
If a central connects to this device, the central subscribes to the Heart Rate Measurement characteristic to receive heart rate notifications forwarded from the peripheral device.
```

> .. _ble_hrs_peripheral_central_sample_testing:

Reviewer 3: this label should be moved before 'Building and running'

Reviewer 2: Good catch. I see the same in the NUS central and HRS central sample READMEs. I'll fix in #799.

User interface
**************

Button 0:
   Press to disable allow list.

   When pairing with authentication, press this button to confirm the passkey shown in the COM listener and complete pairing with the other device.

Button 1:
   Press to disconnect from the connected peripheral device.

   Keep the button pressed while resetting the board to delete bonding information for all peers stored on the device.

   When pairing with authentication, press this button to reject the passkey shown in the COM listener to prevent pairing with the other device.

Button 2:
   Press to disconnect from the connected central device.

LED 0:
   Lit when the device is initialized.

LED 1:
   Lit when connected to a peripheral device.

LED 2:
   Lit when connected to a central device.

Building and running
********************

This sample can be found under :file:`samples/bluetooth/ble_hrs_peripheral_central/` in the |BMshort| folder structure.

For details on how to create, configure, and program a sample, see :ref:`getting_started_with_the_samples`.

Scan filtering options
======================

> The sample always scans for devices advertising the Heart Rate Service UUID

Reviewer 3: ```suggestion
The sample always scans for devices advertising the Heart Rate Service UUID (``0x180D``).
```

Reviewer 2: Fixed

> (``0x180D``). Two optional filters can narrow this down further:

Reviewer 3: ```suggestion
Two optional filters can narrow this down further:
```

Reviewer 2: Fixed.

.. list-table::
   :header-rows: 1

   * - Kconfig option
     - Matches on
   * - :kconfig:option:`CONFIG_SAMPLE_USE_TARGET_PERIPHERAL_NAME`

>      - Advertised device name (eks: ``"MyDeviceName"``)

Reviewer 3: ```suggestion
     - Advertised device name (e.g. ``"MyDeviceName"``)
```

Reviewer 2: Fixed.

   * - :kconfig:option:`CONFIG_SAMPLE_USE_TARGET_PERIPHERAL_ADDR`
     - Exact 48-bit Bluetooth address

Useful for:
* **Development**: Multiple HRS devices nearby, only want to connect to yours.
* **Production**: Bridge paired with one specific sensor, ignore everything else.
* **Bonded scenarios**: Combine with allow lists for tighter control.
Note: "peripheral" in the option name refers to the *remote* device being scanned

> for, not this device's role.

Reviewer: Do we need this?

Reviewer 3: I agree that the bullet points sound a bit like marketing content and don't really bring in new technical information. Please remove the bullets but retain the note, this is actually important.

```suggestion
.. note::

   ``peripheral`` in the option name refers to the *remote* device being scanned for, not this device's role.
```

Reviewer 2: Removed bullet points. Kept the note.

Testing
=======

> This sample can be tested with three devices,

Reviewer 3: ```suggestion

```

Reviewer 2: fixed

> This sample can be tested with three devices, one running this sample, another one running the :ref:`ble_hrs_sample` sample, and a central, for example, a phone or a tablet with `nRF Connect for Mobile`_ or `nRF Toolbox`_.

Reviewer 3: ```suggestion
This sample can be tested with three devices:

* A device running this sample.
* A device running the :ref:`ble_hrs_sample` sample.
* A central device, for example, a phone or a tablet with `nRF Connect for Mobile`_ or `nRF Toolbox`_.
```

Reviewer 2: Done

#. Observe that the ``Scan filter match`` message is printed, followed by ``Connecting to target`` and ``Connected``, when connecting to the peripheral device.
#. Observe that the ``Heart rate service discovered`` message is printed.
#. Connect to the device from nRF Connect (the device is advertising as "nRF_BM_HRS_bridge").
#. Observe that the ``Connecting to target`` and ``Connected`` messages are printed when connecting to the central device.
#. Observe that the device starts receiving heart rate measurement notifications and forwarding them to the central.
#. Note the address printed in the log when connecting, e.g:

   .. code-block:: console

      Connecting to target AA:BB:CC:DD:EE:FF

#. Enable the address filter in Kconfig with:

   .. code-block:: cfg

      CONFIG_SAMPLE_USE_TARGET_PERIPHERAL_ADDR=y
      CONFIG_SAMPLE_TARGET_PERIPHERAL_ADDR=0xAABBCCDDEEFF

   Rebuild and flash. Confirm the sample still connects to the same peripheral.
#. Change the address to a wrong value, rebuild and flash. Confirm the sample no
   longer connects to any peripheral.
#. Repeat the same flow with :kconfig:option:`CONFIG_SAMPLE_USE_TARGET_PERIPHERAL_NAME`

>    (easier, uses the advertised name).

Reviewer: I don't think we should keep this in the test steps, what do you think @anhmolt?

Reviewer 2: Removed this last step.

   .. group-tab:: MCUboot board variants

      The following board variants have DFU capabilities:

      .. include:: /includes/supported_boards_all_mcuboot_variants_s145.txt

Overview
********

This sample advertises and scans for devices that advertise with the :ref:`lib_ble_service_hrs` UUID (0x180D) and initiates a connection when a device is found.
When a peripheral device is connected, the sample starts the service discovery procedure.
If this succeeds, the sample subscribes to the Heart Rate Measurement characteristic to receive heart rate notifications.
If a central connects to this device the central subscribes to the Heart Rate Measurement characteristic to receive heart rate notifications forwarded from the peripheral device.

.. _ble_hrs_peripheral_central_sample_testing:

User interface
**************

Button 0:
  Press to disable allow list.

Button 1:
  Press to disconnect from the connected peripheral device.

  Keep the button pressed while resetting the board to delete bonding information for all peers stored on the device.

Button 2:
  Press to disconnect from the connected central device.

LED 0:
   Lit when the device is initialized.

LED 1:

>    Lit when a peripheral device is connected.

Reviewer 4: ```suggestion
   Lit when connected to a peripheral device.
```
Maybe something like this instead?

LED 2:
   Lit when a central device is connected.

Building and running
********************

This sample can be found under :file:`samples/bluetooth/ble_hrs_peripheral_central/` in the |BMshort| folder structure.

For details on how to create, configure, and program a sample, see :ref:`getting_started_with_the_samples`.

> 

Reviewer 4: ```suggestion
Scan filtering options
**********************

The sample always scans for devices advertising the Heart Rate Service UUID
(``0x180D``). Two optional filters can narrow this down further:

.. list-table::
   :header-rows: 1

   * - Kconfig option
     - Matches on
   * - :kconfig:option:`CONFIG_SAMPLE_USE_TARGET_PERIPHERAL_NAME`
     - Advertised device name (eks: ``"MyDeviceName"``)
   * - :kconfig:option:`CONFIG_SAMPLE_USE_TARGET_PERIPHERAL_ADDR`
     - Exact 48-bit Bluetooth address

Useful for:

* **Development**: Multiple HRS devices nearby, only want to connect to yours.
* **Production**: Bridge paired with one specific sensor, ignore everything else.
* **Bonded scenarios**: Combine with allow lists for tighter control.

Note: "peripheral" in the option name refers to the *remote* device being scanned
for, not this device's role.
```
Adds a new section documenting the optional scan filter Kconfigs, so users know these options exist and what they are for before running the sample.

Complete the following steps to test the sample:

1. Compile and program the application.
#. Observe that the ``BLE HRS Peripheral Central sample initialized`` message is printed.
#. In the Serial Terminal, observe that the ``Advertising as nRF_BM_HRS_bridge`` message is printed.
   You can configure the advertising name using the :kconfig:option:`CONFIG_SAMPLE_BLE_DEVICE_NAME` Kconfig option.
   For information on how to do this, see `Configuring Kconfig`_.
#. Program the second development kit with the :ref:`ble_hrs_sample` sample.
#. Observe that the ``Scan filter match`` message is printed, followed by ``Connecting to target`` and ``Connected``, when connecting to the peripheral device.
#. Observe that the ``Heart rate service discovered.`` message is printed.
#. Connect to the device from nRF Connect (the device is advertising as "nRF_BM_HRS_bridge").
#. Observe that the ``Connecting to target`` and ``Connected`` messages are printed when connecting to the central device.

> #. Observe that the device starts receiving heart rate measurement notifications and forwarding them to the central.

Reviewer 4: ```suggestion
#. Observe that the device starts receiving heart rate measurement notifications and forwarding them to the central.
#. Note the address printed in the log when connecting, eks:

   .. code-block:: console

      Connecting to target AABBCCDDEEFF

#. Enable the address filter in Kconfig with that exact value:

   .. code-block:: cfg

      CONFIG_SAMPLE_USE_TARGET_PERIPHERAL_ADDR=y
      CONFIG_SAMPLE_TARGET_PERIPHERAL_ADDR=0xAABBCCDDEEFF

   Rebuild and flash. Confirm the sample still connects to the same peripheral.

#. Change the address to a wrong value, rebuild and flash. Confirm the sample no
   longer connects to any peripheral.

#. Repeat the same flow with :kconfig:option:`CONFIG_SAMPLE_USE_TARGET_PERIPHERAL_NAME`
   (easier, uses the advertised name).
```
Extend the testing steps to cover the optional scan filters, so users can verify the address and name filtering actually work as expected.
